.travis.yml

```diff
@@ -11,11 +11,204 @@ install:
   - sudo echo "deb [arch=amd64] http://storage.googleapis.com/bazel-apt stable jdk1.8" | sudo tee /etc/apt/sources.list.d/bazel.list
   - curl https://storage.googleapis.com/bazel-apt/doc/apt-key.pub.gpg | sudo apt-key add -
   - sudo apt-get update && sudo apt-get install bazel
+  - sudo apt-get install cppcheck
 
-compiler:
-  - clang
-  - gcc
+matrix:
+  include:
+  # Format
+  - os: linux
+    env:
+      - "TEST=Format"
+    addons:
+      apt:
+        sources:
+          - llvm-toolchain-trusty-4.0
+        packages:
+          - clang-format-4.0
+    script:
+      - clang-format -i tally/include/tally/*.h tally/src/*.{h,cc} tally/tests/*.{h,cc} m3/include/m3/*.h m3/src/*.{h,cc} m3/tests/*.{h,cc}
+      - |
+        if [[ -n $(git diff) ]]; then
+          echo "Source code is not formatted properly"
+          echo ""
+          git diff
+          exit -1
+        fi
 
-script:
-  - bazel build //...
-  # - bazel test //...
+  # CppCheck
+  - os: linux
+    env:
+      - TEST="CppCheck"
+    addons:
+      apt:
+        sources:
+          - ubuntu-toolchain-r-test
+        packages:
+          - gcc-6
+          - g++-6
+    script:
+      - cppcheck --version
+      - cppcheck --inline-suppr --enable=all --error-exitcode=1 -I m3/include/ -I tally/include/ --suppress='*:m3/include/m3/thrift/*' m3/ tally/
+
+  # Clang Tidy
+  - os: linux
+    env:
+      - TEST="ClangTidy"
+    addons:
+      apt:
+        sources:
+          - llvm-toolchain-trusty-4.0
+        packages:
+          - clang-4.0
+          - clang-tidy-4.0
+    script:
+      - export CXX="clang++"
+      - export CC="clang"
+      - bazel build //...
+      - external/compilation-database/generate.sh
+      - clang-tidy -checks='-clang-diagnostic-pragma-once-outside-header' tally/include/tally/*.h tally/src/*.{h,cc}  m3/include/m3/*.h m3/src/*.{h,cc} > output.txt
```

Comment:
Why don't you have a `.clang-tidy` file? I wrote how to generate one somewhere else in this huge review.

Reply:
Ya saw that. I just mean your checks should definitely have a glob star in them or it will use some minimal defaults. `-checks=*,-some-option` should give you many more checks than `-checks=-some-option` IIRC.